internal/common/auth/service_wrapper.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package auth
+
+import (
+    "context"
+    "go.uber.org/cadence/.gen/go/cadence/workflowserviceclient"
+    "go.uber.org/cadence/.gen/go/shared"
+    "go.uber.org/cadence/worker"
+    "go.uber.org/yarpc"
+)
+
+const (
+    jwtHeaderName = "cadence-client-jwt"
+)
+
+type workflowServiceAuthWrapper struct {
+    service workflowserviceclient.Interface
+    authProvider worker.AuthorizationProvider
+}
+
+// NewWorkflowServiceWrapper creates
+func NewWorkflowServiceWrapper(service workflowserviceclient.Interface, authorizationProvider worker.AuthorizationProvider) workflowserviceclient.Interface {
+    return &workflowServiceAuthWrapper{
+        service: service,
+        authProvider: authorizationProvider,
+    }
+}
+
+func (w *workflowServiceAuthWrapper) getYarpcJWTHeader() yarpc.CallOption {
+    token := w.authProvider.GetAuthToken()
+    return yarpc.WithHeader(jwtHeaderName, string(token))
+}
+
+func (w *workflowServiceAuthWrapper) DeprecateDomain(ctx context.Context, request *shared.DeprecateDomainRequest, opts ...yarpc.CallOption) error {
```

Comment:
Hey @longquanzheng , it seems in the metrics implementation of workflowserviceclient, a scope is provided (code [here](https://github.com/uber-go/cadence-client/blob/master/internal/common/metrics/service_wrapper.go#L93)) and the scope is used at every function to handle the possible error (see [here](https://github.com/uber-go/cadence-client/blob/master/internal/common/metrics/service_wrapper.go#L136)). 

The implementation you recommended didn't include a scope

```
service = metrics.NewWorkflowServiceWrapper(service, metricScope)
if authProvider != nil{
    service = auth.NewWorkflowServiceWrapper(service, authProvider)
}
```

So I'm wondering if:

* Either we don't do anything with the scope and we just return the error like this implementation
* We use the same scope as metric
* We create a new one and the new function signate would look like `auth.NewWorkflowServiceWrapper(service, scope, authProvider)`

Also let me know if line 54 looks correct to you

Thanks!

Reply:
It looks good. This wrapper will wrap the wrapper of service metrics so you don’t need to do it again.

Reply:
Actually, we should do this instead:
```
if authProvider != nil{
    service = auth.NewWorkflowServiceWrapper(service, authProvider)
}
service = metrics.NewWorkflowServiceWrapper(service, metricScope)
```
Just a note that this is not overriding the service wrapper with metrics. It takes into a previous one and make another wrapper so the metrics scope is still being used in the previous one. 
Reverting the order here so that the second service warper can capture the erorr of the previous one and emit error metrics. 

Does it make sense?

Reply:
Yeah it makes sense, I need to check the parameters because it's the second time I ask something like this :) Thanks @longquanzheng !

Reply:
Thank you for pointing it out too :D